packages/core/src/common/promise-util.ts

```diff
@@ -59,3 +59,33 @@ export async function retry<T>(task: () => Promise<T>, delay: number, retries: n
 
     throw lastError;
 }
+
+export class AsyncLock {
```

Comment:
Where do you use this class?

Reply:
- https://github.com/eclipse-theia/theia/pull/8546/files#diff-8053c51c6deb015c4dcc93ed38f9daa9R97
- https://github.com/eclipse-theia/theia/pull/8546/files#diff-8053c51c6deb015c4dcc93ed38f9daa9R251

Reply:
I see. Thanks!

From the use-site:
> Ensures that events are processed in the order they are emitted, despite being processed async.

The cool thing is that we do not have to reimplement the weel, you can use a lib. Can `p-queue` help achieve the same?

Reply:
I guess I don't need all the features from `p-queue`, so I saw this implementation I made like writing a for loop: most lines are comments to help someone see through the mechanism. If you tell me that you want me to use `p-queue` I will change it.

Reply:
> I guess I don't need all the features from `p-queue`

You're right, probably you do not need all the features, but it's already in the app no matter you use it or not. 1.2k ⭐ on the GH repo and the 1,788,419 weekly downloads from npmjs suggests it's well-maintained lib that we can use as-is instead of writing additional code.

Reply:
Removed :+1: